Add Dev Services page (#19331)
quarkusio/quarkus: docs/src/main/asciidoc/dev-services.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
////
= Dev Services Overview

> :extension-status: preview

Reviewer: This can be removed.

include::./attributes.adoc[]

Quarkus supports the automatic provisioning of unconfigured services in development and test mode. We refer to this capability
as Dev Services. From a developers perspective this means that if you include an extension and don't configure it then

> Quarkus will automatically start the relevant service (usually using Test Containers behind the scenes) and wire up your

Reviewer: ```suggestion
Quarkus will automatically start the relevant service (usually using https://www.testcontainers.org/[Testcontainers] behind the scenes) and wire up your
```

application to use this service.

All this functionality is part of the Quarkus `deployment` modules, so does not affect the production application in any
way.

This page lists all the Dev Services that Quarkus supports.

> NOTE: In order to use Dev Services you will generally need a working docker environment (remote environments are supported).

Reviewer: ```suggestion
NOTE: In order to use Dev Services you will generally need a working Docker environment (remote environments are supported).
```

> If you don't have docker installed you will need to configure your services normally.

Reviewer: ```suggestion
If you don't have Docker installed you will need to configure your services normally.
```

> == Amqp

Reviewer: ```suggestion
== AMQP
```

> Amqp will be enabled by adding `quarkus-smallrye-reactive-messaging-amqp` to your application.

Reviewer: I'm not sure about the wording here (and in all the other places). It's not clear it's the Dev Service that is enabled. Also I read that as you add the extension to get the dev service but really you add the extension to get the feature and you have the dev service enabled on the side.

Maybe something like:
> The AMQP Dev Service will be enabled when `quarkus-smallrye-reactive-messaging-amqp` is present in your application.

Note that it might just be me not being a native English speaker.

> include::{generated-dir}/config/quarkus-smallrye-reactivemessaging-amqp-config-group-amqp-dev-services-build-time-config.adoc[opts=optional, leveloffset=+1]

Reviewer: We have this page: https://quarkus.io/guides/amqp-dev-services . Apparently, it's not in the index, I will add it.

== Apicurio Registry

Apicurio will be enabled by adding `quarkus-apicurio-registry-avro` to your application.

> include::{generated-dir}/config/quarkus-apicurio-registry-devservices-apicurio-registry-avro-apicurio-registry-dev-services-build-time-config.adoc[opts=optional, leveloffset=+1]

Reviewer: https://quarkus.io/guides/apicurio-registry-dev-services

== Databases

These will be enabled by including a reactive or JDBC datasource extension.

Quarkus provides Dev Services for all databases it supports except Oracle. Most of these are run in a container, with the
exception of H2 and Derby which are run in process. Dev Services are supported for both JDBC and reactive drivers.

> include::{generated-dir}/config/quarkus-datasource-config-group-dev-services-build-time-config.adoc[opts=optional, leveloffset=+1]

Reviewer: Not sure if it's worth linking to https://quarkus.io/guides/datasource#dev-services-configuration-free-databases ?

== Kafka

> include::{generated-dir}/config/quarkus-kafka-client-config-group-kafka-dev-services-build-time-config.adoc[opts=optional, leveloffset=+1]

Reviewer: https://quarkus.io/guides/kafka-dev-services
(also missing from the index, I will add it)

== Keycloak

Keycloak will be enabled by including `quarkus-oidc` to your application.

See link:security-openid-connect-dev-services[OIDC Dev Services]

// currently its not possible to include two different config items with Duration

> // see https://github.com/quarkusio/quarkus/issues/19330

Reviewer: You can add a line with:
```
:no-duration-note: true
```
after the first one.

> // include::{generated-dir}/config/quarkus-keycloak-devservices-keycloak-keycloak-build-time-config.adoc[opts=optional, leveloffset=+1]

Reviewer: https://quarkus.io/guides/security-openid-connect-dev-services (also not in the index)

> == Mongo DB

Reviewer: ```suggestion
== MongoDB
```

> Mongo will be enabled by including `quarkus-mongodb-client` to your application.

Reviewer: ```suggestion
MongoDB will be enabled by including `quarkus-mongodb-client` to your application.
```

> include::{generated-dir}/config/quarkus-mongodb-config-group-dev-services-build-time-config.adoc[opts=optional, leveloffset=+1]

Reviewer: Not much info here: https://quarkus.io/guides/mongodb#dev-services but we should probably add the config properties there too.

== Redis

Redis will be enabled by including `quarkus-redis-client` to your application.

> include::{generated-dir}/config/quarkus-redis-client-config-group-dev-services-config.adoc[opts=optional, leveloffset=+1]

Reviewer: https://quarkus.io/guides/redis-dev-services (also not in the index)

== Vault

Vault will be enabled by including `quarkus-vault` to your application.

> include::{generated-dir}/config/quarkus-vault-config-group-config-dev-services-config.adoc[opts=optional, leveloffset=+1]

Reviewer: https://quarkus.io/guides/vault#dev-services
